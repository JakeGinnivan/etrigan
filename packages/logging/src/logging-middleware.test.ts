import express from 'express'
import supertest from 'supertest'

import { expressRequestLoggingMiddleware, createLogger, scrubJsonLogs } from './'

it('logs requests', async () => {
    const logs: string[] = []
    const originalStdoutWrite = process.stdout.write.bind(process.stdout)
    ;(process.stdout as any).write = (chunk: any, encoding: any, callback: any) => {
        if (typeof chunk === 'string') {
            logs.push(chunk)
        }

        return originalStdoutWrite(chunk, encoding, callback)
    }
    const log = createLogger({
        logLevel: 'debug',
    })
    const middleware = expressRequestLoggingMiddleware(log)
    const app = express()

    app.use(middleware)

    app.get('/', (_req, res) => res.send('Hi'))

    const res = await supertest(app)
        .get('/')
        .expect(200)
    expect(res.text).toBe('Hi')
    expect(scrubJsonLogs(logs)).toMatchSnapshot()
})

it('logs API caller headers', async () => {
    const logs: string[] = []
    const originalStdoutWrite = process.stdout.write.bind(process.stdout)
    ;(process.stdout as any).write = (chunk: any, encoding: any, callback: any) => {
        if (typeof chunk === 'string') {
            logs.push(chunk)
        }

        return originalStdoutWrite(chunk, encoding, callback)
    }
    const log = createLogger({
        logLevel: 'debug',
    })
    const middleware = expressRequestLoggingMiddleware(log)
    const app = express()

    app.use(middleware)

    app.get('/', (_req, res) => res.send('Hi'))

    const res = await supertest(app)
        .get('/')
        .set('Caller', 'custom API caller')
        .set('x-request-id', 'testRequest1')
        .expect(200)
    expect(res.text).toBe('Hi')
    expect(scrubJsonLogs(logs)).toMatchSnapshot()
})

it('logs failed', async () => {
    const logs: string[] = []

    const originalStdoutWrite = process.stdout.write.bind(process.stdout)
    ;(process.stdout as any).write = (chunk: any, encoding: any, callback: any) => {
        if (typeof chunk === 'string') {
            logs.push(chunk)
        }

        return originalStdoutWrite(chunk, encoding, callback)
    }
    const log = createLogger({
        logLevel: 'debug',
    })
    const middleware = expressRequestLoggingMiddleware(log)
    const app = express()

    app.use(middleware)

    app.get('/', (_req, res) => res.status(500).send('Oops'))

    const res = await supertest(app)
        .get('/')
        .expect(500)
    expect(res.text).toBe('Oops')
    expect(scrubJsonLogs(logs)).toMatchSnapshot()
})
