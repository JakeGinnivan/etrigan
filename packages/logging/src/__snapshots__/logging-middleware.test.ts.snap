// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs API caller headers 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"testRequest1\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\",\\"v\\":1}
",
  "{\\"level\\":30,\\"requestId\\":\\"testRequest1\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"testRequest1\\",\\"caller\\":\\"custom API caller\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\",\\"v\\":1}
",
]
`;

exports[`logs close 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\",\\"v\\":1}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request - connection closed\\",\\"v\\":1}
",
]
`;

exports[`logs failed 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\",\\"v\\":1}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\",\\"v\\":1}
",
]
`;

exports[`logs requests 1`] = `
Array [
  "{\\"level\\":20,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"msg\\":\\"begin request\\",\\"v\\":1}
",
  "{\\"level\\":30,\\"requestId\\":\\"<requestid>\\",\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"res\\":{\\"req\\":{\\"query\\":{},\\"requestId\\":\\"<requestid>\\",\\"log\\":{},\\"startTime\\":12345},\\"endTime\\":12345},\\"responseTime\\":0,\\"msg\\":\\"end request\\",\\"v\\":1}
",
]
`;
